ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs

```diff
@@ -264,47 +271,268 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
+      -- block is an EBB and use its block number without subtracting 1.
+      --
+      -- Edge case: if there are two or more consecutive EBBs, we might
+      -- predict the wrong block number, but there are no consecutive EBBs in
+      -- practice, they are one epoch apart.
+    | StoppedAt (HeaderHash blk) BlockNo
+
+      -- | Snoc: the block with the given 'HeaderFields' is in the VolatileDB.
+      -- We also track whether it is an EBB or not.
+      --
+      -- NOTE: we are intentionally lazy in the spine, as constructing the
+      -- path requires lookups in the VolatileDB's in-memory indices, which
+      -- are logarithmic in the size of the index.
+    | (ReversePath blk) ::> (HeaderFields blk, IsEBB)
+
+-- | Lazily compute the 'ReversePath' that starts (i.e., ends) with the given
+-- 'HeaderHash'.
+computeReversePath
+  :: forall blk.
+     LookupBlockInfo blk
+  -> HeaderHash blk
+     -- ^ End hash
+  -> Maybe (ReversePath blk)
+     -- ^ Reverse path from the end point to genesis or the first predecessor
+     -- not in the VolatileDB. Nothing when the end hash is not in the
+     -- VolatileDB.
+computeReversePath lookupBlockInfo endHash =
+    case lookupBlockInfo endHash of
+      Nothing                                               -> Nothing
+      Just blockInfo@VolDB.BlockInfo { bbno, bisEBB, bpreBid } -> Just $
+        go bpreBid bbno bisEBB ::> (headerFieldsFromBlockInfo blockInfo, bisEBB)
+  where
+    go ::
+         WithOrigin (HeaderHash blk)
+         -- ^ The predecessor of the last block added to the path. Not
+         -- necessarily in the VolatileDB.
+      -> BlockNo  -- ^ The block number of the last block
+      -> IsEBB    -- ^ Whether the last block is an EBB or not
+      -> ReversePath blk
+    go predecessor lastBlockNo lastIsEBB = case predecessor of
+      Origin             -> StoppedAtGenesis
+      NotOrigin prevHash -> case lookupBlockInfo prevHash of
+        Nothing ->
+          StoppedAt prevHash (prevBlockNo lastBlockNo lastIsEBB)
+        Just blockInfo@VolDB.BlockInfo { bbno, bisEBB, bpreBid } ->
+          go bpreBid bbno bisEBB ::> (headerFieldsFromBlockInfo blockInfo, bisEBB)
+
+    -- | Predict the block number of the missing predecessor.
+    --
+    -- For regular blocks, this is just block number - 1, EBBs are special of
+    -- course: they share their block number with their predecessor:
+    --
+    -- > block:         regular block 1 | EBB | regular block 2
+    -- > block number:                X |   X | X + 1
+    --
+    -- Edge case: if there are two or more consecutive EBBs, we might predict
+    -- the wrong block number, but there are no consecutive EBBs in practice
+    -- (nor in the tests), they are one epoch apart.
+    prevBlockNo :: BlockNo -> IsEBB -> BlockNo
+    prevBlockNo bno isEBB = case (bno, isEBB) of
+      (0, IsNotEBB) -> error "block with blockNo 0 cannot have a predecessor"
```

Comment:
Byron:
```
Chain:           Genesis :] Block :] Block
Block numbers:   Origin     At 0     At 1
Is EBB:          NA         yes      no
Has predecessor: NA         no       yes
```
Shelley:
```
Chain:           Genesis :] Block :] Block
Block numbers:   Origin     At 0     At 1
Is EBB:          NA         no       no
Has predecessor: NA         no       yes
```
Conclusion: a block with block number 0 is always the very first block and can't have a predecessor. However, that `IsNotEBB` must be removed from that pattern.

Add precondition: must have predecessor.

Rewrite the error message to: "precondition violated".